### ISSUE TITLE
Integrating node-lws into Primus as an available engine

### ISSUE THREAD

**Issue by OtaK** [primus/primus#433]
Hi there guys,

Did you take a look at https://github.com/alexhultman/node-lws ?

This new engine's performance looks purely awesome, is it in the plans to create transports/transformers for it?

Thanks,
Mathieu.

**Comment by lpinca** [comment]
It looks good, but it seems that it doesn't work with an existing node HTTP server (I may be totally wrong).
It definitely needs investigation.

**Comment by lpinca** [comment]
Ref: https://github.com/alexhultman/node-lws/issues/23

**Comment by alexhultman** [comment]
What is primus really? Some kind of standardized wrapper for many different send/receive protocols? Or is it just for WebSockets? What would be required to be included? Is it possible to only implement parts of the interface or are all "transformers" required to fully implement everything?

I have gotten a lot of help from many people and now I'm very close to getting a working http.Server -> node-lws handoff working. It will only support HTTP -> WS for now, but HTTPS -> WSS is something I will look at. Is this something useful for primus?

**Comment by lpinca** [comment]
Hi @alexhultman, Primus is basically a wrapper around different real-time frameworks. It give a common api and allows you to switch the underlying framework (transformer) effortlessly.

It isn't only for WebSocket, we support Engine.IO and SockJS for example and they use different transports (WebSocket, XHR polling, JSONP polling, etc.).

The http.Server -> node-lws handoff is probably all we need to make a working transformer based on node-lws. For example this is how we do it for the websockets transformer: https://github.com/primus/primus/blob/43585c2f14221fd68f13728aca73b943e39a2545/transformers/websockets/server.js#L37-L3.

It should support HTTPS->WSS but having only HTTP->WS would be a good starting point as I guess it would be possible to use a ssl terminating reverse proxy.

**Comment by alexhultman** [comment]
I don't know what a real-time framework is but hey - let's do it. There will be initial shortcomings but I think primus inntegration would highlight eventual interface issues / lacking features in node-lws so it could be a nice project. Node-lws is only for servers currently and I haven't really planned any client support.

**Comment by alexhultman** [comment]
I'm trying to run the echo example but I'm getting ENOENT because something wants to read primus/dist/primus.js when I connect. Am I supposed to build something to get that dist folder because I don't have it.

```
Open http://localhost:8080 in your browser
fs.js:549
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: ENOENT: no such file or directory, open '/home/alexhultman/primus/dist/primus.js'
    at Error (native)
    at Object.fs.openSync (fs.js:549:18)
    at Object.fs.readFileSync (fs.js:397:15)
    at Primus.read (/home/alexhultman/primus/index.js:164:24)
    at Primus.compile (/home/alexhultman/primus/index.js:508:20)
    at Primus.client (/home/alexhultman/primus/middleware/primus.js:30:44)
    at iterate (/home/alexhultman/primus/transformer.js:131:20)
    at done (/home/alexhultman/primus/transformer.js:139:7)
    at Primus.control (/home/alexhultman/node_modules/access-control/index.js:91:39)
    at iterate (/home/alexhultman/primus/transformer.js:136:14)
```

Also, how is the compatibility between transformers? Will it be possible to run one transformer on the server but another on the client as long as the transport is the same? Like, transports are protocols and transformers are implementations of that protocol right?

**Comment by lpinca** [comment]
Yes, run `npm run build` to build the client. This will put it under `dist/primus.js`.
You should be able to use the client of the already existing `websockets` or `faye` transformers as a client for this transformer.

**Comment by alexhultman** [comment]
Omg I typed ```npm build```, soooo close!

**Comment by alexhultman** [comment]
This is where I'm at:

```
'use strict';

var http = require('http')
  , parse = require('url').parse;

/**
 * Minimum viable WebSocket server for Node.js that works through the Primus
 * interface.
 *
 * @runat server
 * @api private
 */
module.exports = function server() {
  var WebSocketServer = require('lws').Server
    , logger = this.logger
    , Spark = this.Spark;

  var service = this.service = new WebSocketServer({
    perMessageDeflate: !!this.primus.options.compression
    /*clientTracking: false,
    noServer: true*/
  });

  var pendingUpgrades = {};

  service.on('message', function (socket, message, binary) {
    var spark = service.getUserData(socket);
    spark.emits('incoming::data')(message.toString()); // I would rather leave it as a Node.js buffer
  });

  service.on('close', function (socket) {
    var spark = service.getUserData(socket);
    spark.someDisconnectionMark = true;
    spark.emits('incoming::end')(); // yes?
  });

  service.on('connection', function (socket) {
    var req = pendingUpgrades[0]; // placeholder
    var spark = new Spark(
        req.headers               // HTTP request headers.
      , req                       // IP address location.
      , parse(req.url).query      // Optional query string.
      , null                                    // We don't have an unique id.
      , req                       // Reference to the HTTP req.
    );

    service.setUserData(socket, spark);

    spark.on('outgoing::end', function end() {
      //if (socket) socket.close();
    }).on('outgoing::data', function write(data) {

      // Ignore sends on closed sockets
      if (spark.someDisconnectionMark) {
        return;
      }

      if ('string' === typeof data) {  
        service.send(socket, new Buffer(data), false);
      } else {
        service.send(socket, data, true); // I guess we get Node.js buffers in this case?
      }
    });

    // spark.emits('incoming::error')
    // spark.emits('incoming::ping')
  });

  /**
   * Noop! Pointless, empty function that will actually be really useful.
   *
   * @param {Error} err We failed at something.
   * @api private
   */
  function noop(err) {
    if (err) logger.error(err);
  }

  //
  // Listen to upgrade requests.
  //
  this.on('upgrade', function upgrade(req, socket, head) {
    pendingUpgrades[0] = req;
    this.service.handleUpgrade(socket, req, head);
  });

  //
  // Listen to non-upgrade requests.
  //
  this.on('request', function request(req, res) {
    res.writeHead(426, { 'content-type': 'text/plain' });
    res.end(http.STATUS_CODES[426]);
  });

  this.on('close', function close() {
    service.close(); // this is not supported
  });
};
```

It works with the echo server but as you can see I have just guessed the most part.

**Comment by alexhultman** [comment]
I've done some improvements to the interface to make this primus thing easier:

```javascript
'use strict';

var http = require('http')
  , parse = require('url').parse;

/**
 * Minimum viable WebSocket server for Node.js that works through the Primus
 * interface.
 *
 * @runat server
 * @api private
 */
module.exports = function server() {
  var WebSocketServer = require('/home/alexhultman/node-lws/dist/lws').Server
    , logger = this.logger
    , Spark = this.Spark;

  var service = this.service = new WebSocketServer({
    perMessageDeflate: !!this.primus.options.compression
  });

  service.on('message', function (socket, message, binary) {
    var spark = service.getUserData(socket);
    if (binary) {
      spark.emits('incoming::data')(message);
    } else {
      spark.emits('incoming::data')(message.toString()); // ?
    }
  });

  service.on('close', function (socket) {
    var spark = service.getUserData(socket);
    spark.emits('incoming::end')(); // ?
  });

  /**
   * Noop! Pointless, empty function that will actually be really useful.
   *
   * @param {Error} err We failed at something.
   * @api private
   */
  function noop(err) {
    if (err) logger.error(err);
  }

  //
  // Listen to upgrade requests.
  //
  this.on('upgrade', function upgrade(req, socket, head) {
    var socket = this.service.handleUpgrade(socket, req, head);

    var spark = new Spark(
        req.headers               // HTTP request headers.
      , req                       // IP address location.
      , parse(req.url).query      // Optional query string.
      , null                      // We don't have an unique id.
      , req                       // Reference to the HTTP req.
    );

    service.setUserData(socket, spark);

    spark.on('outgoing::end', function end() {
      //if (socket) socket.close();

      // i need to check the spark status, ther is no check on the socket
      service.close(socket);
    }).on('outgoing::data', function write(data) {

      if (Spark.CLOSED === this.readyState) {
        return;
      }

      if ('string' === typeof data) {  
        service.send(socket, new Buffer(data), false);
      } else {
        service.send(socket, data, true);
      }
    });

    // spark.emits('incoming::error')
    // spark.emits('incoming::ping')
  });

  //
  // Listen to non-upgrade requests.
  //
  this.on('request', function request(req, res) {
    res.writeHead(426, { 'content-type': 'text/plain' });
    res.end(http.STATUS_CODES[426]);
  });

  this.on('close', function close() {
    service.close(); // this is not supported
  });
};
```

So you get the basic idea: from 'upgrade' you handleUpgrade and get a socket and there you create the spark and attach the spark to the socket. Then you get the spark from the 'message' and 'close' events. I would appreciate some info on how I should call the primus things. Like, is this somewhat correct?

**Comment by lpinca** [comment]
@alexhultman thanks, overall it seems ok. It needs a little cleanup but I understand your logic here.
Do you mean how you should call this transformer or what are the primus things?

**Comment by alexhultman** [comment]
Hehe, I mean - how am I supposed to use the spark? Am I doing it correctly? I cannot find any documentation listing the function signatures of the events and what kind of types they take, etc (do they take Node.js buffers, strings or what do they take).

A socket in node-lws has no status, so I was planning on using the attached spark to track the status of the socket. Can I read (and set) readyState for this purpose? And I can check it against Spark.CLOSED etc?

Is this correct: spark.emits('incoming::end')(); ? Does it properly change the spark's readyState? I need it to change so that I won't send on a closed socket.

Also, what's up with setting the HTTP headers in the spark? Will these get deleted later on? Otherwise that would add a ton of overhead per connection. Are they optional to set?

Anyways - thanks for the feedback.

**Comment by alexhultman** [comment]
Ah, I looked some more in spark.js and it seems the spark automatically tracks the status and discards invalid sends and such. I guess this should be correct then?

```javascript
'use strict';

var http = require('http')
  , parse = require('url').parse;

/**
 * Minimum viable WebSocket server for Node.js that works through the Primus
 * interface.
 *
 * @runat server
 * @api private
 */
module.exports = function server() {
  var WebSocketServer = require('lws').Server
    , logger = this.logger
    , Spark = this.Spark;

  var service = this.service = new WebSocketServer({
    perMessageDeflate: !!this.primus.options.compression
  });

  service.on('message', function (socket, message, binary) {
    var spark = service.getUserData(socket);
    if (binary) {
      spark.emits('incoming::data')(message);
    } else {
      spark.emits('incoming::data')(message.toString());
    }
  });

  service.on('close', function (socket) {
    var spark = service.getUserData(socket);
    spark.emits('incoming::end')();
  });

  /**
   * Noop! Pointless, empty function that will actually be really useful.
   *
   * @param {Error} err We failed at something.
   * @api private
   */
  function noop(err) {
    if (err) logger.error(err);
  }

  //
  // Listen to upgrade requests.
  //
  this.on('upgrade', function upgrade(req, socket, head) {
    var socket = this.service.handleUpgrade(socket, req, head);
    if (socket !== undefined) {
      var spark = new Spark(
          req.headers               // HTTP request headers.
        , req                       // IP address location.
        , parse(req.url).query      // Optional query string.
        , null                      // We don't have an unique id.
        , req                       // Reference to the HTTP req.
      );

      service.setUserData(socket, spark);

      spark.on('outgoing::end', function end() {
        service.close(socket);
      }).on('outgoing::data', function write(data) {
        if ('string' === typeof data) {
          // todo: take a string directly
          service.send(socket, new Buffer(data), false);
        } else {
          service.send(socket, data, true);
        }
      });
    }
  });

  //
  // Listen to non-upgrade requests.
  //
  this.on('request', function request(req, res) {
    res.writeHead(426, { 'content-type': 'text/plain' });
    res.end(http.STATUS_CODES[426]);
  });

  this.on('close', function close() {
    service.close();
  });
};
```

**Comment by lpinca** [comment]
Shipped with `primus@5.0.0`, thank you @alexhultman.
